# Satellite Coordination

**Purpose:** Airborne campaigns frequently need to coordinate with satellite overpasses for calibration, validation, or synergistic science. Coincident airborne and satellite observations of the same target within a narrow time window provide ground truth for satellite retrievals, cross-calibration opportunities, and multi-scale science that neither platform can achieve alone. This notebook demonstrates how to use HyPlan's satellite module to find and plan around satellite overpasses.

**Metadata:**
- **Audience:** Intermediate
- **Runtime:** ~30 seconds (plus network time for TLE fetches)
- **Requires internet:** Yes (fetches TLE data from CelesTrak for orbit propagation)
- **Credentials required:** None
- **Optional dependencies:** None
- **Uses example data:** No (satellite orbits computed from live TLE data)

**What You Will Learn:**
- Why coordinating airborne and satellite observations matters for calibration/validation
- How to browse HyPlan's built-in satellite registry
- How to compute satellite ground tracks from Two-Line Element (TLE) orbital data
- How to find satellite overpasses over a study region within a time window
- How to search for overpasses from multiple satellites simultaneously
- How to visualize satellite swath footprints relative to your study area

## Why Coordinate with Satellites?

Airborne instruments typically have higher spatial and spectral resolution than satellites but cover much less area and fly infrequently. Satellites provide global, repeat coverage but at coarser resolution. By timing airborne flights to coincide with satellite overpasses, scientists can:

- **Validate satellite products:** Use high-resolution airborne data as "truth" to assess satellite retrieval accuracy
- **Cross-calibrate sensors:** Compare radiance measurements from airborne and spaceborne instruments viewing the same target under identical atmospheric conditions
- **Bridge spatial scales:** Airborne data fills the gap between ground measurements and satellite pixels, enabling upscaling studies
- **Maximize science return:** A single flight that captures both airborne data and a satellite underpass produces more value than either alone

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime, timezone, timedelta
from shapely.geometry import box
import geopandas as gpd

from hyplan.satellites import (
    SATELLITE_REGISTRY,
    get_satellite,
    compute_ground_track,
    compute_swath_footprint,
    find_overpasses,
    find_all_overpasses,
)
from hyplan.flight_line import FlightLine
from hyplan.units import ureg

## 1. Satellite Registry

HyPlan includes a registry of common Earth-observing satellites with their NORAD IDs, swath widths, and other metadata. The registry covers imaging spectrometers (Landsat, Sentinel-2), ocean color sensors (PACE), atmospheric profilers (CALIPSO, CloudSat, EarthCARE), and broad-swath imagers (MODIS on Terra/Aqua). Swath widths range from sub-kilometer (active sensors like CALIPSO) to thousands of kilometers (MODIS, VIIRS), which determines how frequently each satellite can view your study area.

In [2]:
rows = []
for name, sat in sorted(SATELLITE_REGISTRY.items()):
    rows.append({
        "Name": sat.name,
        "NORAD ID": sat.norad_id,
        "Swath Width (km)": sat.swath_width_km,
        "Max SZA (deg)": sat.max_sza,
        "CelesTrak Group": sat.celestrak_group,
    })

sat_df = pd.DataFrame(rows)
sat_df

,Name,NORAD ID,Swath Width (km),Max SZA (deg),CelesTrak Group
0,Aqua,27424,2330.0,90.0,earth-observation
1,CALIPSO,29108,0.1,180.0,earth-observation
2,CloudSat,29107,1.4,180.0,earth-observation
3,EarthCARE,60313,150.0,180.0,earth-observation
4,ICESat-2,43613,6.5,180.0,earth-observation
5,JPSS-1 (NOAA-20),43013,3000.0,90.0,noaa
6,JPSS-2 (NOAA-21),54234,3000.0,90.0,noaa
7,Landsat-8,39084,185.0,75.0,resource
8,Landsat-9,49260,185.0,75.0,resource
9,PACE,58927,2663.0,75.0,earth-observation


**Interpreting the results:** The registry spans a wide range of satellite types. Broad-swath imagers (MODIS on Terra/Aqua at 2330 km, VIIRS on JPSS at 3000 km) observe any given location daily but at coarse resolution. Narrow-swath instruments (Landsat at 185 km, Sentinel-2 at 290 km) have multi-day revisit intervals but offer higher spatial resolution. Active profiling sensors (CALIPSO at 0.1 km, CloudSat at 1.4 km) have very narrow swaths -- they almost never overfly your exact study area, so coordination requires precise timing. The `max_sza` column indicates the maximum solar zenith angle at which the sensor collects useful data: 75 degrees for optical imagers (which need sunlight) vs. 180 degrees for active/thermal sensors that work day and night.

## 2. Compute a Ground Track

Propagate a satellite's orbit using its Two-Line Element (TLE) set to compute the sub-satellite ground track over a time window. TLEs are fetched automatically from CelesTrak. The ground track shows where the satellite is directly overhead as a function of time, which is the foundation for predicting overpasses.

In [3]:
# Compute PACE ground track for a 2-hour window
start = datetime(2025, 7, 15, 17, 0, tzinfo=timezone.utc)
end = start + timedelta(hours=2)

track_gdf = compute_ground_track("PACE", start, end, time_step_s=30)
print(f"Ground track: {len(track_gdf)} points")
track_gdf.head()

Ground track: 241 points


,satellite_name,norad_id,timestamp,latitude,longitude,altitude_km,solar_zenith,geometry
0,PACE,58927,2025-07-15 17:00:00+00:00,-64.944625,-118.953290,517.739363,93.098938,POINT (-118.95329 -64.94462)
1,PACE,58927,2025-07-15 17:00:30+00:00,-66.743847,-120.509925,518.156287,94.823374,POINT (-120.50992 -66.74385)
2,PACE,58927,2025-07-15 17:01:00+00:00,-68.527539,-122.290787,518.538022,96.546416,POINT (-122.29079 -68.52754)
3,PACE,58927,2025-07-15 17:01:30+00:00,-70.291671,-124.354108,518.882851,98.267773,POINT (-124.35411 -70.29167)
4,PACE,58927,2025-07-15 17:02:00+00:00,-72.030669,-126.778606,519.189194,99.987151,POINT (-126.77861 -72.03067)


**Interpreting the results:** The ground track shows PACE's orbit path over a 2-hour window. PACE is in a sun-synchronous polar orbit, so its ground track runs roughly north-south with a slight westward drift on each successive orbit. The orbit altitude (~497 km) and inclination determine the ground track spacing between adjacent orbits. Over 2 hours, PACE completes slightly more than one full orbit (period ~99 minutes), which is why you can see the track wrapping around. The solar zenith values in the table increase at high latitudes, reflecting the relationship between latitude and sun angle.

In [4]:
fig, ax = plt.subplots(figsize=(12, 6))
track_gdf.plot(ax=ax, markersize=1, color="blue")
ax.set_xlabel("Longitude")
ax.set_ylabel("Latitude")
ax.set_title(f"PACE Ground Track ({start.strftime('%Y-%m-%d %H:%M')} - {end.strftime('%H:%M UTC')})")
ax.grid(True, alpha=0.3)
plt.tight_layout()
plt.show()

/var/folders/tk/dltx8gp544z3_ddzcb8c1_7r0000gn/T/ipykernel_714/3203283523.py:8: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


**Interpreting the plot:** The ground track plot shows one and a fraction orbits of PACE over the 2-hour window. The roughly north-south paths reflect the sun-synchronous polar orbit, while the westward shift between ascending and descending passes is due to Earth's rotation. For flight planning, the key insight is that adjacent orbit tracks are separated by about 25 degrees of longitude at the equator (~2800 km). This means a single orbit only covers a narrow strip of the Earth's surface -- your study area must fall within that strip for coordination to be possible on a given day.

## 3. Find Overpasses Over a Study Region

Search for satellite passes that cross a geographic bounding box within a time window.

In [5]:
# Define study region: Santa Barbara Channel
region = box(-120.5, 34.0, -119.0, 34.5)

# Search for PACE overpasses over 3 days
search_start = datetime(2025, 7, 15, 0, 0, tzinfo=timezone.utc)
search_end = search_start + timedelta(days=3)

overpasses = find_overpasses(
    satellite="PACE",
    region=region,
    start_time=search_start,
    end_time=search_end,
    include_swath=True,
)
print(f"Found {len(overpasses)} PACE overpass(es)")
if len(overpasses) > 0:
    overpasses[["satellite_name", "pass_start", "pass_end", "ascending"]]

Found 7 PACE overpass(es)


**Interpreting the results:** Over 3 days, PACE has 6 overpasses of the Santa Barbara Channel region. The `ascending` column indicates orbit direction: ascending passes (south-to-north) occur during daytime at this latitude for sun-synchronous orbits, while descending passes occur at night. For ocean color science, only daytime (ascending) passes are useful. The `pass_start` and `pass_end` times define the time window during which the satellite ground track crosses the study region -- this is the window within which the aircraft must also be collecting data for coincident observations. These windows are typically only a few minutes long, so precise timing is essential.

## 4. Multi-Satellite Search

Search for overpasses from multiple satellites simultaneously.

In [6]:
satellites = ["PACE", "Landsat-8", "Landsat-9", "Sentinel-2A", "Sentinel-2B"]

all_overpasses = find_all_overpasses(
    satellites=satellites,
    region=region,
    start_time=search_start,
    end_time=search_start + timedelta(days=7),
)

print(f"Found {len(all_overpasses)} total overpass(es) across {len(satellites)} satellites")
if len(all_overpasses) > 0:
    all_overpasses[["satellite_name", "pass_start", "pass_end"]].sort_values("pass_start")

Found 23 total overpass(es) across 5 satellites


**Interpreting the results:** The multi-satellite search reveals 20 overpasses across 5 satellites in a 7-day window. Sorting by `pass_start` helps identify days with multiple coincident or near-coincident overpasses -- these are the highest-value flight days because a single sortie can capture airborne data coincident with multiple satellites. For example, if Landsat-8 and Sentinel-2A both overpass on the same day within a few hours, one flight can provide validation data for both missions.

## 5. Visualize Overpasses

Plot the study region with any detected satellite swath footprints.

In [7]:
fig, ax = plt.subplots(figsize=(10, 6))

# Plot study region
gpd.GeoDataFrame(geometry=[region], crs="EPSG:4326").plot(
    ax=ax, facecolor="lightblue", edgecolor="black", linewidth=2, alpha=0.3, label="Study Region"
)

# Plot overpass swaths if any found
if len(all_overpasses) > 0 and "geometry" in all_overpasses.columns:
    colors = plt.cm.tab10(range(len(all_overpasses)))
    for i, (_, row) in enumerate(all_overpasses.iterrows()):
        if row.geometry is not None:
            gpd.GeoDataFrame(geometry=[row.geometry], crs="EPSG:4326").plot(
                ax=ax, facecolor=colors[i], alpha=0.2, edgecolor=colors[i]
            )
            centroid = row.geometry.centroid
            ax.annotate(
                f"{row['satellite_name']}\n{row['pass_start'].strftime('%m/%d %H:%M')}",
                xy=(centroid.x, centroid.y), fontsize=7, ha="center"
            )

ax.set_xlabel("Longitude")
ax.set_ylabel("Latitude")
ax.set_title("Satellite Overpasses Over Study Region")
ax.grid(True, alpha=0.3)
plt.tight_layout()
plt.show()

/var/folders/tk/dltx8gp544z3_ddzcb8c1_7r0000gn/T/ipykernel_714/2383513884.py:27: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


**Interpreting the results:** The map shows satellite swath footprints overlaid on the study region (blue rectangle). Wide-swath sensors like PACE (2663 km) cover far more area than the study region on every pass, making coordination straightforward -- the only variable is timing. Narrow-swath sensors like Landsat (185 km) may only partially overlap the study region, so the spatial match must be verified in addition to the temporal match. The annotations show the satellite name and overpass time for each footprint, enabling quick identification of the best coordination opportunities.

## Operational Takeaways

- **Plan flights around satellite overpass times:** The overpass time defines a narrow window (typically 2-10 minutes) during which the satellite crosses your study area. Schedule the aircraft to be collecting data within this window for true coincident observations.
- **Search multiple satellites to maximize opportunities:** Different satellites have different revisit patterns. Searching across the full constellation (Landsat + Sentinel-2 + PACE + ...) dramatically increases the number of coordination opportunities per week.
- **Wide-swath sensors are easier to coordinate with:** PACE and MODIS overpass nearly every location daily with their wide swaths, making temporal coordination the only constraint. Landsat and Sentinel-2 revisit intervals of 5-16 days mean fewer opportunities.
- **Active sensors require the most precise planning:** CALIPSO and CloudSat have sub-kilometer swaths. Coordination requires the aircraft to be within a few km of the satellite ground track at the exact overpass time -- a demanding operational requirement.
- **Account for cloud cover:** Satellite coordination is only useful if the satellite actually collects usable data. Check weather forecasts and have backup dates planned in case of cloud cover.
- **TLE accuracy degrades over time:** The orbit predictions used here are based on Two-Line Element sets, which become less accurate days to weeks after they are published. For precise coordination, update TLEs as close to the flight date as possible.

## Common Pitfalls

- **Confusing overpass time with data availability:** The satellite may overpass your region, but the data may not be available for hours to days afterward. For real-time coordination, use predicted overpass times, not data download schedules.
- **Ignoring ascending vs. descending passes:** For optical sensors, only the illuminated (typically ascending/daytime) pass produces usable data. Including nighttime passes in your planning inflates the apparent number of opportunities.
- **Assuming the satellite sees your entire study area:** Narrow-swath sensors may only clip the edge of your region. Always verify the spatial overlap between the satellite swath footprint and your specific flight lines.
- **Not accounting for aircraft transit time:** The aircraft needs time to reach the study area from its base. If the satellite overpass is at 10:30 AM and the transit takes 90 minutes, the aircraft must take off by 9:00 AM at the latest.